client/src/main/java/org/spine3/type/TypeUrl.java

```diff
@@ -116,35 +125,45 @@ public static TypeUrl from(EnumDescriptor descriptor) {
     }
 
     /**
-     * Creates a new instance from the passed type URL or type name.
+     * Creates a new instance from the passed {@code TypeName}.
+     */
+    static TypeUrl from(TypeName typeName) {
+        checkNotNull(typeName);
+        final TypeUrl typeUrl = KnownTypes.getTypeUrl(typeName.value());
+        return typeUrl;
+    }
+
+    /**
+     * Creates a new instance from the passed type URL.
      *
-     * @param typeUrlOrName the type URL of the Protobuf message type or its fully-qualified name
+     * @param typeUrl the type URL of the Protobuf message type
      */
     @Internal
-    public static TypeUrl of(String typeUrlOrName) {
-        checkNotEmptyOrBlank(typeUrlOrName, "type URL or name");
-        final TypeUrl typeUrl = isTypeUrl(typeUrlOrName)
-                                ? ofTypeUrl(typeUrlOrName)
-                                : ofTypeName(typeUrlOrName);
-        return typeUrl;
+    public static TypeUrl parse(String typeUrl) {
+        checkNotNull(typeUrl);
+        checkArgument(!typeUrl.isEmpty());
+        checkArgument(isTypeUrl(typeUrl), "Malformed type URL: %s", typeUrl);
+
+        final TypeUrl result = doParse(typeUrl);
+        return result;
     }
 
     private static boolean isTypeUrl(String str) {
         return str.contains(SEPARATOR);
     }
 
-    private static TypeUrl ofTypeUrl(String typeUrl) {
+    private static TypeUrl doParse(String typeUrl) {
         final String[] parts = typeUrlSeparatorPattern.split(typeUrl);
         if (parts.length != 2) {
             throw malformedTypeUrl(typeUrl);
         }
 
-        final String prefix = parts[0].trim();
+        final String prefix = parts[0];
         if (prefix.isEmpty()) {
             throw malformedTypeUrl(typeUrl);
         }
 
-        final String typeName = parts[1].trim();
+        final String typeName = parts[1];
```

Comment:
Why don't we `trim` the parts anymore? "` `" part also indicates a malformed URL.